tools/data_gen/src/main.rs

```diff
@@ -0,0 +1,197 @@
+#[macro_use]
+extern crate lazy_static;
+
+use argh::FromArgs;
+use calyx::{errors::CalyxResult, frontend, ir};
+use serde::ser::{Serialize, Serializer};
+use serde_json::{json, Map, Value};
+use std::borrow::Borrow;
+use std::cell::{Ref, RefCell};
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::rc::Rc;
+
+lazy_static! {
+    static ref SIZEMAP: HashMap<&'static str, Vec<&'static str>> = {
+        vec![
+            ("std_mem_d1", vec!["SIZE"]),
+            ("std_mem_d2", vec!["D0_SIZE", "D1_SIZE"]),
+            ("std_mem_d3", vec!["D0_SIZE", "D1_SIZE", "D2_SIZE"]),
+            (
+                "std_mem_d4",
+                vec!["D0_SIZE", "D1_SIZE", "D2_SIZE", "D3_SIZE"],
+            ),
+        ]
+        .into_iter()
+        .collect::<HashMap<&'static str, Vec<&'static str>>>()
+    };
+}
+
+///Holds data for std_mem cells, including name of cell, width, and sizes
+///Name is the name of cell itself, not its type. Sizes is a vector
+/// that holds the dimensions of the cell (ex: for a 2 x 3 std_mem_d2 cell it would be [2,3])
+struct CellData {
+    name: String,
+    width: u64,
+    sizes: Vec<u64>,
+}
+
+enum SizeVec {
+    Quad(Vec<Vec<Vec<Vec<u64>>>>),
+    Triple(Vec<Vec<Vec<u64>>>),
+    Double(Vec<Vec<u64>>),
+    Single(Vec<u64>),
+}
+
+impl Serialize for SizeVec {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match &*self {
+            SizeVec::Single(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 3, "data", &v)
+            }
+            SizeVec::Double(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 2, "data", &v)
+            }
+            SizeVec::Triple(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 1, "data", &v)
+            }
+            SizeVec::Quad(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 0, "data", &v)
+            }
+        }
+    }
+}
```

Comment:
I may be mistaken, but I think serde can derive this particular serialization implementation. Does `#[derive(Serialize)` not do the right thing here?

Reply:
#derive[(Serialize)] runs fine, except it prints out the json like this:
"data": {"Single": [data in here]} when I want
"data": [data in here]
I was able to eventually get what I wanted using #derive[(Serialize)] through this line in rust:
`json!({"data": data.as_object().unwrap_or_else(|| panic!("couldn't place data array into json")).values().next_back()})`
There may be a more optimal way to do it though.

Reply:
Sounds like you want an untagged enum (see https://serde.rs/enum-representations.html#untagged)